feat(deepagents): add python documentation (#957)
langchain-ai/docs: src/oss/deepagents/customization.mdx

---

> title: Customizing Deep Agents

Reviewer: is "deep agents" a product name? or just a type of agent? if not a product, this should not be capitalized

description: Learn how to customize deep agents with system prompts, tools, subagents, and more
---

## System Prompt

Deep agents come with a built-in system prompt that is heavily based on and inspired by Claude Code's system prompt. The default system prompt contains detailed instructions for how to use the built-in planning tool, file system tools, and sub agents.

> Each deep agent tailored to a use case should include a custom system prompt specific to that use case as well!

Reviewer: ```suggestion
Each deep agent tailored to a use case should include a custom system prompt specific to that use case.
```
Combine w above paragraph

    query: str,
    max_results: int = 5,
    topic: Literal["general", "news", "finance"] = "general",
    include_raw_content: bool = False,
):
    """Run a web search"""
    return tavily_client.search(
        query,
        max_results=max_results,
        include_raw_content=include_raw_content,
        topic=topic,
    )

agent = create_deep_agent(
    model="anthropic:claude-sonnet-4-20250514",
    tools=[internet_search]
)
```

In addition to any tools that you provide, deep agents also get access to a number of default tools. These are:

- **write_todos** - Update the agent's to-do list
- **ls** - List all files in the agent's filesystem
- **read_file** - Read a file from the agent's filesystem
- **write_file** - Write a new file in the agent's filesystem
- **edit_file** - Edit an existing file in the agent's filesystem
- **task** - Spawn a subagent to handle a specific task

You can learn more about these tools in the [PlanningMiddleware](/oss/deepagents/middleware#planning-middleware), [FilesystemMiddleware](/oss/deepagents/middleware#filesystem-middleware), and [SubAgentMiddleware](/oss/deepagents/middleware#subagent-middleware) sections!

## Subagents

> A main feature of Deep Agents is their ability to spawn subagents. You can specify custom subagents that your agent can hand off work to in the `subagents` parameter. Sub agents are useful for [context quarantine](https://www.dbreunig.com/2025/06/26/how-to-fix-your-context.html#context-quarantine) (to help not pollute the overall context of the main agent) as well as custom instructions.

Reviewer: ```suggestion
A main feature of deep agents is their ability to create subagents. You can specify custom subagents that your agent can hand off work to in the `subagents` parameter. Subagents are useful for [context quarantine](https://www.dbreunig.com/2025/06/26/how-to-fix-your-context.html#context-quarantine) (to help not pollute the overall context of the main agent) as well as custom instructions.
```

```python
from deepagents import create_deep_agent, CompiledSubAgent
from langchain.agents import create_agent
from langgraph.pregel.remote import RemoteGraph

# Create a custom agent graph
custom_graph = create_agent(
    model=your_model,
    tools=specialized_tools,
    prompt="You are a specialized agent for data analysis..."
)

# Use it as a custom subagent
custom_subagent = CompiledSubAgent(
    name="data-analyzer",
    description="Specialized agent for complex data analysis tasks",
    runnable=custom_graph
)

subagents = [custom_subagent]

agent = create_deep_agent(
    model="anthropic:claude-sonnet-4-20250514",
    tools=[internet_search],
    system_prompt=research_instructions,
    subagents=subagents
)
```

In addition to any user defined subagents, deep agents have access to a `general-purpose` subagent at all times - this is a subagent with the same instructions as the main agent and all the tools that is has access to. The primary purpose of the `general-purpose` subagent is to context isolation. The main agent can always delegate a complex task to this subagent, and get a concise answer back without any bloat from intermediate tool calls.

> ## Human-in-the-Loop

Reviewer: ```suggestion
## Human-in-the-loop
```

```python
from langchain_core.tools import tool
from deepagents import create_deep_agent
from langgraph.store.memory import InMemoryStore

@tool
def get_weather(city: str) -> str:
    """Get the weather in a city."""
    return f"The weather in {city} is sunny."

store = InMemoryStore()  # Or any other Store object
agent = create_deep_agent(
    model="anthropic:claude-sonnet-4-20250514",
    tools=[get_weather],
    store=store,
    use_longterm_memory=True
)
```

Extending a deep agent with long-term memory allows all of the filesystem tools to access and store files in the long-term memory store. Specifically, any files that start with the path `/memories/` are saved in long-term memory. This store can be accessed **across** different threads.

An example usage of long-term memory is to store instructions for the agent. An agent can be prompted to call its `edit_file` tool to update its own instructions file in `/memories/instructions.txt` any time a user provides some feedback. Overtime, the instructions.txt file in long-term memory accumulates more and more of the users preferences, so that the agent can get better at its job.

> ## Additional Middleware

Reviewer: ```suggestion
## Additional middleware
```
